torch/utils/data/datapipes/iter/combinatorics.py

```diff
@@ -74,19 +76,29 @@ class ShuffleIterDataPipe(IterDataPipe[T_co]):
     def __init__(self,
                  datapipe: IterDataPipe[T_co],
                  *,
-                 buffer_size: int) -> None:
+                 buffer_size: int = 10000,
+                 unbatch_level: int = 0
+                 ) -> None:
         super().__init__()
         assert buffer_size > 0, "buffer_size should be larger than 0"
-        self.datapipe = datapipe
+        if unbatch_level == 0:
+            self.datapipe = datapipe
+        else:
+            self.datapipe = datapipe.unbatch(unbatch_level=unbatch_level)
         self.buffer_size = buffer_size
         self._buffer = []
+                
+    def buffer_replace(self, x):
+        idx = random.randint(0, self.buffer_size - 1)
+        val = self._buffer[idx]
+        self._buffer[idx] = x
+        return val
 
     def __iter__(self) -> Iterator[T_co]:
+        # TODO: Buffer is global, should be per __iter__ !!!
```

Comment:
Is this TODO aiming to make shuffle shardable?

Reply:
It is just standard behaviour of iterators to be independent from each other